src/racer/util.rs

```diff
@@ -79,6 +79,87 @@ pub fn symbol_matches(stype: SearchType, searchstr: &str, candidate: &str) -> bo
     }
 }
 
+pub fn find_closure(src: &str) -> Option<(ByteRange, ByteRange)> {
+    let (pipe_range, _) = closure_valid_arg_scope(src)?;
+    let mut chars = src.chars()
+                       .enumerate()
+                       .skip(pipe_range.end.0)
+                       .skip_while(|(_, c)| {
+                           c.is_whitespace()
+                       });
+    let (start, start_char) = if let Some((i, c)) = chars.next() {
+        (if c == '{' { i + 1 } else { i }, c)
+    } else {
+        return None;
+    };
+    let (mut clevel, mut plevel) = if start_char == '{' {
+        (1i32, 0i32)
+    } else {
+        (0i32, 0i32)
+    };
+    let mut chars = chars.peekable();
+    while chars.peek().is_some() {
+        let (i, current) = chars.next().unwrap();
+        match current {
+            '{' => clevel += 1,
+            '(' => plevel += 1,
+            '}' => {
+                clevel -= 1;
+                if clevel == 0 && (start_char == '{') {
+                    return Some((pipe_range, ByteRange::new(BytePos(start), BytePos(i))));
+                }
+            },
+            ';' => {
+                if start_char != '{' {
+                    return Some((pipe_range, ByteRange::new(BytePos(start), BytePos(i))));
+                }
+            },
+            ')' => {
+                plevel -= 1;
+                if plevel == -1 {
+                    return Some((pipe_range, ByteRange::new(BytePos(start), BytePos(i))));
+                }
+                if plevel == 0 && start_char != '{' {
+                    match chars.peek() {
+                        Some((_, c)) if *c == '.' => {}
```

Comment:
I don't like this heuristics since it has many edge cases(e.g. `a.method()[] or a.method() + 5` is also can be available.)
How about just waiting `;` or `plevel == -1`